### ISSUE TITLE
[NOT AN ISSUE] Training for RTL languages and scripts

### ISSUE THREAD

**Issue by johnlockejrr** [rvankoert/loghi#23]
Hi there!

Can I train a recogniton model for RTL languages (Hebrew, Arabic or Syriac)? If so, should I do something specific or the normal preocedure will be ok? I already have ground truth/datasets in `PAGE-XML` format.

Thank you so much!

**Comment by rvankoert** [comment]
For the HTR part: It should work as bi-directional lstm is used by default, but the padding is by default on the right which might cause problems. We have never tried RTL languages and there is no switch in the command line to inverse the direction. I'm happy to add this switch + functionality if you could share some of your ground truth with me or point me to an RTL dataset for verification purposes.
For the layout(laypa): there should be no problem, but the default models will have been trained on LTR material.
For the loghi-tooling: there might be problems as we expect LTR, but I can't think of any right now.

Once again: If you want we can try to include RTL-support as it seems a minimal effort.

**Comment by johnlockejrr** [comment]
Thank you so much for replying and your kind openness!
I have ground truth in `PAGE-XML` format, if is ok I can send you the link in email, the dataset is not yet relased for the public.
Here is my email: jognlockejrr [ at ] gmail dot com.
Or if the `PyLaia` format dataset is ok for the recognition model I have it on HuggingFace.
Hebrew dataset: https://huggingface.co/datasets/johnlockejrr/sofer_mahir
Samaritan dataset: https://huggingface.co/datasets/johnlockejrr/samaritan_v1
For segmentation I also have one in YOLO segmentation format: https://huggingface.co/datasets/johnlockejrr/sam_seg_yolo
Let me know if is ok.
Thank you!

**Comment by rvankoert** [comment]
Hi johnlockejrr,

I sent you an email and removed your address from your message.

**Comment by johnlockejrr** [comment]
Good.

By the way, latest PyLaia and Kraken use python-bidi for this, they don't
flip the image.

**Comment by rvankoert** [comment]
ah, after reading py-bidi again I understand it now. Swapping that in

**Comment by rvankoert** [comment]
ok, thanks for the hint :) it's now 
--bidirectional to enable bidi
It works nicely and does not seem to interfere with the flow of latin texts. Most of the testing is finished and looks good. Some other unrelated code needs to come along with this release and that is still running its tests.

**Comment by rvankoert** [comment]
the code is in the main repo and is included in release 2.2.15. That release is slightly delayed due to network issues.

**Comment by johnlockejrr** [comment]
Perfect! Can't wait to test it!

**Comment by johnlockejrr** [comment]
Are the `latest` docker images updated to `2.2.15` release? I pulled them and edited the `./scripts/inference-pipeline.sh` to `VERSION=latest`

**Comment by rvankoert** [comment]
yes, it is in 2.2.15 The dockers are available. 
To train with RTL/bidi-support add in the training command:
--bidirectional
In the scripts folder there is a training example (htr-train-pipeline.sh) . Just add it somewhere between lines 98 and 110.

There are more options explained at [https://github.com/knaw-huc/loghi-htr](https://github.com/knaw-huc/loghi-htr) . 
May I suggest some augmentation options to add also in the training script:
--aug_elastic_transform
--aug_random_crop
--aug_random_width
--aug_random_shear
--aug_distort_jpeg
I'm in the process of simplifying/completing the documentation, but to be honest, that is usually last on my todo-list.

**Comment by johnlockejrr** [comment]
Thank you so much for clarifying!
To train a segmentation model on `laypa` I should stick to the default?
I mean, there are some differences between RTL and LTR on page layout and text line. Could the htr models take care of that?

**Comment by rvankoert** [comment]
I think it is best to stick to the defaults with laypa. For laypa I don't think LTR or RTL would matter on how you train it. Laypa has quite good documentation. The HTR models should take care of everything if i'm correct. We make one assumption in the loghi-tooling: text has a somewhat horizontal  orientation. Baselines are extracted as LTR from the laypa output (same as the dataset you sent me), but in the HTR itself it is then fixed if run with --bidirectional 

@stefanklut : do you have any suggestions for laypa and RTL scripts?

**Comment by stefanklut** [comment]
I think there won't be any changes needed, considering LTR or RLT will give the same baseline. It's the loghi-tooling which then cuts it out based on the predicted baseline. If you have GT with the baseline going from RTL in pageXML on transkribus, I believe that would mean the baseline is upside down. But again, the pixels for such a baseline are identical. I only see an issue with both upside down and RTL text being mixed that may cause issue. Even if laypa isn't the best on upside down text to start with. Of course different scripts (e.g. Hebrew, Arabic) will have different looking text, so I do suggest training (or finetuning) the model to better differentiate between text and other black parts of a scan (bleed through, smudges)
